satpy/readers/seviri_l2_bufr.py

```diff
@@ -0,0 +1,225 @@
+        fh = open(self.filename)
+
+        bufr = ec.codes_bufr_new_from_file(fh)
+        ec.codes_set(bufr, 'unpack', 1)
+
+        key_arr = []
+        iterid = ec.codes_bufr_keys_iterator_new(bufr)
+        while ec.codes_bufr_keys_iterator_next(iterid):
+            key_arr.append(ec.codes_bufr_keys_iterator_get_name(iterid))
+        ec.codes_bufr_keys_iterator_delete(iterid)
+        ec.codes_release(bufr)
+
+        fh.close()
+        return key_arr
+
+    @property
+    def start_time(self):
+        return self.rc_start
+
+    @property
+    def end_time(self):
+        return self.rc_start+timedelta(minutes=15)
+
+    def get_array(self, parameter, mnbr=None):
+        ''' Get ancilliary array data eg latitudes and longitudes '''
+
+        with open(self.filename) as fh:
+            fh = open(self.filename)
+
+            if mnbr is not None:
+                bufr = ec.codes_bufr_new_from_file(fh)
+                ec.codes_set(bufr, 'unpack', 1)
+                arr = ec.codes_get_array(bufr, parameter)
+                ec.codes_release(bufr)
+
+            else:
+                msgCount = 0
+                while True:
+
+                    bufr = ec.codes_bufr_new_from_file(fh)
+                    if bufr is None:
+                        break
+
+                    ec.codes_set(bufr, 'unpack', 1)
+                    # if is the first message initialise our final array with
+                    # the number of subsets contained in the first message
+                    if (msgCount == 0):
+                        arr = np.zeros(ec.codes_get(bufr, 'numberOfSubsets'), np.float)
+                        arr[:] = ec.codes_get_array(bufr, parameter, float)
+
+                    else:
+                        tmpArr = np.zeros(ec.codes_get(bufr, 'numberOfSubsets'), np.float)
+                        tmpArr[:] = ec.codes_get_array(bufr, parameter, float)
+                        arr = np.concatenate((arr, tmpArr))
+
+                    msgCount = msgCount+1
+                    ec.codes_release(bufr)
+
+        fh.close()
+        if arr.size == 1:
+            arr = arr[0]
+
+        return arr
+
+    def get_attribute(self, parameter, mnbr=None):
+        ''' Get BUFR attributes '''
+
+        fh = open(self.filename, "rb")
+
+        if mnbr is not None:
+            bufr = ec.codes_bufr_new_from_file(fh)
+            ec.codes_set(bufr, 'unpack', 1)
+            arr = ec.codes_get_array(bufr, parameter)
+            ec.codes_release(bufr)
+
+        fh.close()
+
+        if arr.size == 1:
+            arr = arr[0]
+
+        return arr
+
+    def get_dataset(self, dsid, info):
+        ''' here we loop through the BUFR file and for the required key
+        append the data vlues to the dataset area'''
+
+        arr2 = np.empty((self.Ny, self.Nx)).astype(np.float)
+        arr2.fill(np.nan)
+
+        with open(self.filename, "rb") as fh:
+            fh = open(self.filename, "rb")
```

Comment:
why do you need to open the file twice ?

Reply:
error in my part - resolved, only 1 open i made